### DISCUSSION TITLE
Tree modeling with dynamic types (by example of JSON)

### DISCUSSION THREAD

**Discussion by fungs** [capnproto/capnproto#1880] [General]
Hello everyone,

I want to understand the different ways how to model a tree with static and dynamic type definitions (I'm using the Python bindings at the moment, so some practical limitations may apply). This can be useful for a wide variety of applications. 

**Example**

Let's assume we want to simplistically model JSON primitives to create a dynamic JSON binary representation with arbitrary JSON objects at runtime (schemaless aka self-describing).

    struct Field {
        union {
            string @0 :StringField;
            number @1 :NumberField;
            bool @2 :BoolField;
            array @3 :ArrayField;
            object @4 :ObjectField;
            null @5 :Void;
        }
    }
    
    using StringField = Text;
    using NumberField = Int64;
    using BoolField = Bool;
    using ArrayField = List(Field);
    using ObjectField = List(ObjectKV);  # root of data structure
    
    struct ObjectKV {
        key @0 :Text;
        value @1 :Field;
    }

**Assumptions**

Here are my basic assumptions. Please correct me if you think I'm wrong.

1. The model can be used to serialize and deserialize JSON-encoded data.
2. An encoded `Field` carries a union marker to distinguish the basic types with extra space requirements.

**Questions**

1. Is there a maybe more efficient schemaless model (possibly involving generics)?
2. Say we want to save bandwidth or space, strip the union marker, and instead keep an external JSON schema for decoding. What would be a good way to encode the tree under these circumstances? My obvious takes would be to remove the union and supply type information for the fields at deserialization time. How can we construct and cast the objects at runtime (using Python)?
3. If we are able to do (2), is there any advantage in adding a uniform list type, e.g. `List[Int64]` which is standard in most typed languages and large data applications?
4. Assuming that the same JSON structure be used every time, we could convert the JSON schema itself into CAPNP representation, basically replacing JSON objects in the model by native structs and using a fixed tree topology instead of a dynamic one. What would be the advantages in terms of compute and space?

**Final remark**

I've successfully used CAPNP in Python to serialize dynamic large data. The example seems trivial, but zero copy, in-place construction and random access reader already provide a huge benefit working with large data. Exploring this use case will help me and others to make best use of CAPNP in these applications. Thanks for your contributions!

**Comment by kentonv** [comment]
Check out json.capnp, which actually uses a schema very similar to what you've written: https://github.com/capnproto/capnproto/blob/v2/c++/src/capnp/compat/json.capnp

> Is there a maybe more efficient schemaless model (possibly involving generics)?

Well... You could always encode actual JSON/BSON/CBOR/MsgPack/etc. into a `Data` field.

But if you want to keep the type represented as much as possible in Cap'n Proto's type system, then not really. I think this is sort of a pathological case: Cap'n Proto is good at data that follows a schema, but here you're trying to encode schemaless data, so you're creating a meta-schema, but at this point Cap'n Proto encoding is losing a lot of its benefits and instead adding bloat to the encoding.

> Say we want to save bandwidth or space, strip the union marker, and instead keep an external JSON schema for decoding. What would be a good way to encode the tree under these circumstances? My obvious takes would be to remove the union and supply type information for the fields at deserialization time. How can we construct and cast the objects at runtime (using Python)?

You could abandon the schema entirely and instead use the `AnyStruct` API to manually encode your `Field` struct. You can then manually specify the sizes of the data section and pointer section. So if the value is a number, you allocate your struct to contain only a data section. If it's a string, you have only a pointer section.

You could also drop the field names and instead use an array of values that corresponds to the ordering of the fields in the JSON schema.

But again at this point I'm not sure you're getting much benefit from Cap'n Proto, and it might just be better to use some other encoding stuffed into a `Data` field.

> If we are able to do (2), is there any advantage in adding a uniform list type, e.g. List[Int64] which is standard in most typed languages and large data applications?

Sorry, I don't understand this question. Cap'n Proto already has `List(Int64)`.

> Assuming that the same JSON structure be used every time, we could convert the JSON schema itself into CAPNP representation, basically replacing JSON objects in the model by native structs and using a fixed tree topology instead of a dynamic one. What would be the advantages in terms of compute and space?

This is what [Cap'n Proto's JSON library](https://github.com/capnproto/capnproto/blob/v2/c++/src/capnp/compat/json.h) does. You can define a Cap'n Proto schema corresponding to some JSON schema, and then parse directly into the Cap'n Proto structure / serialize a Cap'n Proto structure to JSON. The main advantage over typical JSON parsers is that you get to use the type-safe classes generated by Cap'n Proto to manipulate your JSON data. Additionally, the Cap'n Proto encoding of those structures (if you choose to use it) may be smaller than the JSON encoding, due to not having to include field names -- though the Cap'n Proto encoding does have a lot more padding which in some cases can make it larger overall.

Note that the current JSON implementation is not well-optimized, particularly for parsing.

Also note that I don't think the JSON library is currently exposed in pycapnp, but I suspect it wouldn't be hard to add.

**Comment by fungs** [comment]
Thanks, @kentonv, for the detailed answer and the pointers to JSON-related code!

For better understanding of the use case: I'm actually exploring a mixed case for large data in which parts of the message are statically typed while other parts of the message change at runtime, with their own, application-specific, external schemas (including extended validation syntax and extensible types). I can basically see four usable options to handle those dynamic parts in a message:

1. Opaque (with external schema): encoding as (single) binary blob, and delegating the handling to a different serialization protocol entirely.
2. Dynamic topology in CAPNP (with external schema): encoding a tree structure without type information (as binary), e.g. using the array-like, ordered representation of class-like objects without field names.
3. Dynamic topology and types in CAPNP (schemaless): encoding a tree with custom type information (as in the JSON example above).
4. Static topology and types in CAPNP (external schema only needed for features unavailable in CAPNP schema): encoding everything in CAPNP style, by using on-the-fly CAPNP schema generation by some sort of translation layer

My understanding for those options is:

1. has the issue that most serialization protocols are not designed for large data
2. seems a reasonable approach to me, basically creating a hierarchical, indexed, binary value store, allowing selective deserialization with custom protocols for structured data, using random access and without having to transport the type information on the wire
3. works as seen above, but adds quite some overhead due to redundant type encoding in the union, in particular, if uniform typed arrays are encoded via JSON style lists, which add the type markers for every element instead of once (thus my related question)
4. adds builtin type validation and tooling and probably has the equal or smaller overhead compared to option 2, I need to implement the translation layer and additional validation on top of this.

----------

>> You could abandon the schema entirely and instead use the AnyStruct API to manually encode your Field struct. You can then manually specify the sizes of the data section and pointer section. So if the value is a number, you allocate your struct to contain only a data section. If it's a string, you have only a pointer section.

What I was after for option 2 was something like

    struct Field(T) {
      data @0: T;
    }

and then doing something like `Field<Bool>(data=true).cast_as(Field<T>)` on serialization and `field.cast_as(Field<Bool>).data` on deserialization. I believe this is effectively what you are suggesting with the AnyStruct API only that I don't know how to do it in Python in either case. Does anyone know how to do it and show an example?

----------

>> If we are able to do (2), is there any advantage in adding a uniform list type, e.g. List[Int64] which is standard in most typed languages and large data applications?

> Sorry, I don't understand this question. Cap'n Proto already has List(Int64).

I meant adding the CAPNP type `List(Int64)` as a custom type to the shown JSON model. This relates to option 3, and making it more efficient in encoding for uniform array data.

Let me reformulate my question: How much more overhead has `List[Field]` compared to `List[NumberType]` (obviously all entries are numbers)?

And on top, relating to option 2 vs. 4: Do `List[Data]` and `List[Int64]` have the same space requirements on the wire?

Thanks for your input!

**Reply by kentonv** [reply]
> parts of the message are statically typed while other parts of the message change at runtime, with their own, application-specific, external schemas (including extended validation syntax and extensible types).

Can the applications use Cap'n Proto for their app-specific schemas? If they can be Cap'n Proto, then I think that's the cleanest answer by far. It's easy to embed "Cap'n Proto structure of unknown schema" into your message using the `AnyPointer` type, or generics. Moreover, Cap'n Proto schemas can be encoded in their own capnp-based format (`schema.capnp`) from which they can be loaded dynamically using Cap'n Proto's schema API (`capnp/schema.h`). You can then use the dynamically-loaded schema to manipulate the application's data in a dynamic way using Cap'n Proto's dynamic API (`capnp/dynamic.h`).

Of course, a requirement to use Cap'n Proto may or may not be acceptable to your application developers...

> What I was after for option 2 was something like

Note that Cap'n Proto generics only work over pointer types, not primitives. That is, you can't specialize `Field(T)` as `Field(Bool)`, because `Bool` is a primitive type. You would need to define a boxed boolean type (`struct BoolBox { value @0 :Bool; }`) and then use that (`Field(BoolBox)`). This is inefficient, of course, since the `Field` has to contain an 8-byte pointer to a separate 8-byte object which only really contains 1 bit of data. With `AnyStruct`, you would be able to remove a layer of indirection and encode the boolean directly in the data section of `Field` -- but this is a very low-level interface, and you'd have to manually decide how to encode a bool (or any other non-pointer type) into bytes.

(The reason capnp generics have this limitation is so that all specializations of `Field` have the same layout, since all pointers are the same size. If different specializations had different layouts, it would add an unreasonable amount of complexity to the implementation.)

> Let me reformulate my question: How much more overhead has List[Field] compared to List[NumberType] (obviously all entries are numbers)?

This really depends on the exact definitions of `Field` and `NumberType`. As it turns out, `List(BoxedUInt64)` and `List(UInt64)` have the same encoding. But the answer changes if `NumberType` is smaller than 8 bytes, or if `Field` contains a pointer to the value rather than embedding the value directly.

It might be worth reading the encoding docs so you can answer these questions directly: https://capnproto.org/encoding.html

**Comment by fungs** [comment]
Thanks again, I will go through the docs and clarify my requirements based on your answers. The only thing I'm still missing is a statement about added extra bits the union adds to each field (it might be in the encoding docs you referenced, in which case I will find it!

I know are probably not the right person to address for Python questions, but I still don't know how the AnyPointer API can be used in Python.

**Reply by kentonv** [reply]
A union tag is always 16 bits, as if you added a `:UInt16` field to the struct. The overall struct size is rounded up to a multiple of 8 bytes, though, and since your `Field` struct includes an `Int64` field, it was already perfectly at an 8-byte boundary without the tag... so the tag ends up adding 8 bytes (2 bytes value, 6 bytes padding).

**Reply by fungs** [reply]
Great, that was the kind of information I was hoping for! It helps me to consider the optimal encoding.
